app/controllers/split_checkout_controller.rb

```diff
@@ -96,11 +100,23 @@ def update_order
 
     @order.select_shipping_method(params[:shipping_method_id])
     @order.update(order_params)
+    # We need to update voucher to take into account:
+    #  * when moving away from "details" step : potential change in shipping method fees
+    #  * when moving away from "payment" step : payment fees
+    recalculate_voucher if details_step? || payment_step?
     @order.update_totals_and_states
 
     validate_current_step
   end
 
+  def recalculate_voucher
+    return if @order.voucher_adjustments.empty?
+
+    return if @order.shipping_method&.id == params[:shipping_method_id].to_i
```

Comment:
I'm still not familiar with vouchers yet.
I have a few questions:

1. why don't we use the same guard closure with payment_method?
2. what will happen if the `VoucherAdjustmentsService` is called and the `shipping_method` isn't changed?

Reply:
> * why don't we use the same guard closure with payment_method?

You are right we should, I'll have a look at that. As explained below, the guard check is there to avoid doing unnecessary processing.

> * what will happen if the `VoucherAdjustmentsService` is called and the `shipping_method` isn't changed?

It would just recalculated the voucher adjustment and end up with the same result, I added this mostly to avoid doing processing that is not necessary.

Reply:
I looked at adding a guard clause for payment_method, but it adds a lots complexity with the current code, which could probably be alleviated by refactoring split checkout controller but I don't think it's worth the effort. In most cases we want to recalculate the voucher when leaving the payment step, but not when leaving the details step. So I think it's fine here to not have guard clause. 

I really want to refactor the split checkout so it's clearer what's happening on each step, but I don't want to invest the time now. There is further work to be done on vouchers so maybe an opportunity will arise.